File: apps/ai-app/app/services/greeting.py
from app.core.logging import get_logger
from app.core.response import ErrorResponse, SuccessResponse, success_response
from app.core.trace import get_tracer
from app.repository.openai.dependency import DepGreetingRepoOpenAI

tracer = get_tracer("service.greetings")


class GreetingService:
    __log = get_logger()

    def __init__(self, repo: DepGreetingRepoOpenAI):
        self.__repo = repo

    async def greetings(self) -> SuccessResponse[dict] | ErrorResponse:
        with tracer.start_as_current_span("service.greetings") as span:
            self.__log.info("Service layer log", extra={"layer": "service"})
            greetings = await self.__repo.greetings()
            return success_response(data=greetings)

File: templates/rust-ai/src/services/greeting.rs
//! Greeting service — orchestration layer.
//!
//! Mirrors `app/services/greeting.py::GreetingService`. Holds an
//! `Arc<GreetingRepoOpenAI>` and exposes a single async method
//! `greetings()` that wraps the repository response in the standard
//! `SuccessResponse` envelope.

use std::sync::Arc;

use axum::response::Response;
use tracing::Instrument;

use crate::core::exception::AppError;
use crate::core::response::success_response;
use crate::repository::openai::greeting::GreetingRepoOpenAI;

pub struct GreetingService {
    repo: Arc<GreetingRepoOpenAI>,
}

impl GreetingService {
    pub fn new(repo: Arc<GreetingRepoOpenAI>) -> Self {
        Self { repo }
    }

    /// REPLICABLE span: `service.greetings`. Must exact-match the Python
    /// `with tracer.start_as_current_span("service.greetings")` call.
    pub async fn greetings(&self) -> Result<Response, AppError> {
        let span = tracing::info_span!("service.greetings");
        async {
            tracing::info!(layer = "service", "Service layer log");
            let data = self.repo.greetings().await?;
            Ok(success_response(data))
        }
        .instrument(span)
        .await
    }
}
